File: code/day8.R
parse_boot <- function(file) {
  strsplit(readLines(file), " ")
}

boot <- parse_boot("data/day8")

# puzzle 1
accum <- 0
instr <- c()
line <- 1
seen <- rep(FALSE, length(boot))
while (line <= length(boot)) {
  if (seen[line] == TRUE) {
    res <- list(instr, accum)
    break
  }
  
  if (boot[[line]][1] == "nop") {
    seen[line] <- TRUE
    instr <- c(instr, line)
    line <- line + 1
  } else if (boot[[line]][1] == "acc") {
    seen[line] = TRUE
    accum <- accum + as.integer(boot[[line]][2])
    instr <- c(instr, line)
    line <- line + 1
  } else if (boot[[line]][1] == "jmp") {
    seen[line] <- TRUE
    instr <- c(instr, line)
    line <- line + as.integer(boot[[line]][2])
  }
}

# puzzle 2
for (i in res[[1]]) {
  bootcopy <- boot
  if (bootcopy[[i]][1] == "nop") {
    bootcopy[[i]][1] <- "jmp"
  } else if (boot[[i]][1] == "jmp") {
    bootcopy[[i]][1] <- "nop"
  } else {
    next
  }
  
  accum <- 0
  line <- 1
  seen <- rep(FALSE, length(boot))
  fixed <- "yes"
  while (line <= length(boot)) {
    if (seen[line] == TRUE) {
      fixed <- "nope"
      break
    }
    
    if (bootcopy[[line]][1] == "nop") {
      seen[line] <- TRUE
      line <- line + 1
    } else if (bootcopy[[line]][1] == "acc") {
      seen[line] = TRUE
      accum <- accum + as.integer(bootcopy[[line]][2])
      line <- line + 1
    } else if (bootcopy[[line]][1] == "jmp") {
      seen[line] <- TRUE
      line <- line + as.integer(bootcopy[[line]][2])
    }
  }
  
  if (fixed == "yes") {
    break
  }
}

File: code/day8.jl
function program(code)
    line = 1
    acc = 0
    log = Set()
    infinite = false
    while line <= size(code, 1)
        if line ∈ log
            infinite = true
            break
        end
        push!(log, line)
        inst, val = split(code[line], " ")
        if inst == "nop"
            line += 1
        elseif inst == "acc"
            line += 1
            acc += parse(Int, val)
        elseif inst == "jmp"
            line += parse(Int, val)
        end
    end
    acc, infinite, log
end

function fixcode(code)
    broke = program(code)
    for line in broke[3]
        codecopy = copy(code)
        inst, val = split(codecopy[line], " ")
        if inst == "acc"
            continue
        elseif inst == "jmp"
            codecopy[line] = "nop " * val
        elseif inst == "nop"
            codecopy[line] = "jmp " * val
        end
        result = program(codecopy)
        if result[2] == false
            return result[1]
            break
        end
    end
end

code = readlines("data/day8")
# puzzle 1
program(code)

# puzzle 2
fixcode(code)
